Firestore/core/src/local/query_engine.h

```diff
@@ -30,44 +30,84 @@ enum class LimitType;
 namespace local {
 
 class LocalDocumentsView;
+class IndexManager;
 
+// TODO(cheryllin): Add function name into documentation which configures index
+// (e.g. setIndexConfiguration).
 /**
- * A query engine that takes advantage of the target document mapping in the
- * TargetCache. Query execution is optimizes by only reading the documents that
- * previously matched a query plus any documents that were edited after the
- * query was last listened to.
+ * Firestore queries can be executed in three modes. The Query Engine determines
+ * what mode to use based on what data is persisted. The mode only determines
+ * the runtime complexity of the query - the result set is equivalent across all
+ * implementations.
  *
- * There are some cases where Index-Free queries are not guaranteed to
- * produce the same results as full collection scans. In these cases, the
- * query processing falls back to full scans. These cases are:
+ * The Query engine will use indexed-based execution if a user has configured
+ * any index that can be used to execute query. Otherwise, the engine will try
+ * to optimize the query by re-using a previously persisted query result. If
+ * that is not possible, the query will be executed via a full collection scan.
  *
- * - Limit queries where a document that matched the query previously no
- *   longer matches the query.
+ * Index-based execution is the default when available. The query engine
+ * supports partial indexed execution and merges the result from the index
+ * lookup with documents that have not yet been indexed. The index evaluation
+ * matches the backend's format and as such, the SDK can use indexing for all
+ * queries that the backend supports.
+ *
+ * If no index exists, the query engine tries to take advantage of the target
+ * document mapping in the TargetCache. These mappings exists for all queries
+ * that have been synced with the backend at least once and allow the query
+ * engine to only read documents that previously matched a query plus any
+ * documents that were edited after the query was last listened to.
+ *
+ * There are some cases where this specific optimization is not guaranteed to
+ * produce the same results as full collection scans. In these cases, query
+ * processing falls back to full scans. These cases are:
+ *
+ * - Limit queries where a document that matched the query previously no longer
+ * matches the query.
  * - Limit queries where a document edit may cause the document to sort below
- *   another document that is in the local cache.
+ * another document that is in the local cache.
```

Comment:
This part of comment should be updated, we no longer resort to full collection scan for the first two cases, we run the query without limit using index now.

Reply:
I remove the first two cases in the coming commit. Just two questions:

1. It is appropriate to add something like: `limit query will be run without limit using index` ? Would this lead to bill confusion?

2. Do I need to update Android comments as well?